### PR TITLE
[TensorExpr] Cache use of fallback in kernel invocation

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48264 [NNC] Preserve strided output
* #48263 Remove inferred from tensor type ctors
* #47814 [TensorExpr Fuser] Add support for nodes which have tensor constant inputs
* #47813 [NNC] Compute Tensor Output Properties in ininitialization
* **#47812 [TensorExpr] Cache use of fallback in kernel invocation**
* #47810 Dont use symbolic shapes check

Summary:

Previously we were checking the environment every kernel invocation for `tensorExprFuserEnabled`, which checks the environment for `PYTORCH_TENSOREXPR`. This is only a dev-exposed API, so I think it is fine to only check once when the kernel is initialized. The `disable_optimization` flag which is user-exposed more or less covers the same functionality. 

For fun, some benchmarking. I compared scripted before and after of
```
def foo(x, y):
    return x + y
```
for x, y = torch.tensor([1]). I also removed the prim::TypeCheck node to better
isolate the kernel (I cheated). Here is gist: https://gist.github.com/eellison/39f3bc368f5bd1f25ded4827feecd15e

Without Changes Run 1:
no fusion: sum 6.416894399004377 min: 0.6101883250012179 median 0.6412974080012646
with fusion: sum 6.437897570998757 min: 0.6350401220006461 median 0.6446951820034883

Without Changes Run2:
no fusion: sum 6.601341788002173 min: 0.6292048720024468 median 0.6642187059987918
with fusion: sum 6.734651455997664 min: 0.6365462899993872 median 0.6755226659988693

With Changes Run1:
no fusion: sum 6.097717430002376 min: 0.5977709550024883 median 0.613631643998815
with fusion: sum 6.1299369639964425 min: 0.5857932209983119 median 0.6159247440009494

With Changes Run2:
no fusion: sum 6.5672018059995025 min: 0.6245676209982776 median 0.6386050750006689
with fusion: sum 6.489086147994385 min: 0.6236886289989343 median 0.6535737619997235

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25286210](https://our.internmc.facebook.com/intern/diff/D25286210)

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -1853,36 +1853,35 @@ void TensorExprKernel::compile() {
...
     : graph_(subgraph), code_(subgraph, "") {
-  if (!fallbackAllowed()) {
+
+  allow_fallback_ = fallbackAllowed();
+  if (!allow_fallback_) {
     compile();
     return;
   }
 
+  use_fallback_ = fallbackEnforced();
```
Comment (Krovatkin) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1863]: nitpick: maybe renaming `use_fallback_` to `failed_comp_or_required_fallback_` makes it a bit more obvious in which cases we are using the fallback?

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 2]
```diff
@@ -1853,36 +1853,35 @@ void TensorExprKernel::compile() {
...
-  if (fallback_) {
-    fallback(stack);
-    return;
-  }
-  try {
+  if (!use_fallback_ && !allow_fallback_) {
     runKernel(stack);
-  } catch (...) {
-    fallback_ = true;
+  } else if (!use_fallback_ && allow_fallback_) {
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1878]: I was trying to optimize for the common case - i guess try catch doesn't have a pereformance penalty if exceptions aren;t thrown though (?)

**Review by bertmaher — APPROVED** [review]
Is fallback actually useful anymore?  Since we default it to off should we consider killing it?

**Comment by protonu** [comment]
> Is fallback actually useful anymore? Since we default it to off should we consider killing it?

I think I use the fallback for Block Codegen. I can look into removing the dependency on fallback.

**Pull request closed without merging**